### PR TITLE
I18n "in" step for Custom Blocks

### PR DESCRIPTION
**Pull request by joshlory**
First step towards custom block redaction.  Ingest the source yaml file from individual block JSON files.  My plan is to finish redaction/restoration plugins, then re-enable `blocks.en.yml` in Crowdin.

Sample `blocks.en.yml` content:

```yml
---
en:
  data:
    blocks:
      Dancelab_addBehavior:
        text: sprite {SPRITE} begins {BEHAVIOR}
      Dancelab_atTimestamp:
        text: at {TIMESTAMP} seconds
      Dancelab_changeColorBy:
        text: change {COLOR} {METHOD} by {AMOUNT}
        options:
          '"hue"': hue
          '"saturation"': saturation
          '"brightness"': brightness
      Dancelab_changeMove:
        text: change {SPRITE} dance move to {MOVE}
        options:
          '0': Standing
          '1': Clap
          '2': Spin
          '3': Uprock
          '4': Bboy
          '"next"': Next
          '"rand"': Random
```

### CODE REVIEW

**Comment by joshlory** [comment]
[Implement plugin for redacting & restoring JSON custom block definitions](https://trello.com/c/CSzP5g36/238-implement-plugin-for-redacting-restoring-json-custom-block-definitions)

File: bin/i18n/sync-codeorg-in.rb [hunk 1]
```diff
@@ -32,6 +33,26 @@ def sanitize(string)
...
+  Dir.glob('dashboard/config/blocks/**/*.json').sort.each do |file|
+    name = File.basename(file, '.*')
+    config = JSON.parse(File.read(file))['config']
+    blocks[name] = {
+      'text' => config['blockText'],
+    }
+
+    next unless config['args']
+    args_with_options = config['args'].map {|arg| arg['options']}.compact.flatten(1)
+    blocks[name]['options'] = args_with_options.map(&:reverse).to_h unless args_with_options.empty?
```
Comment (Hamms) [inline bin/i18n/sync-codeorg-in.rb]: is this going to be sufficient information to consume these strings at the other end? Seems like we would want this to be something like `blocks[name]['options'][arg['name']]`

Comment (joshlory) [reply bin/i18n/sync-codeorg-in.rb]: I'm mixing all the options together into a single hash.  So if two dropdowns in a block both have "north" as a key, they'll have to translate the same way in both dropdowns.  But a different block could have "north" in a dropdown translate to something else.

Does that seem like the right level of granularity?  I could make the keys more specific, so each option is grouped under a key for the name of the arg (what you said).  Or I could make the keys even less specific, so "north" as a dropdown option anywhere always translates to the same string.

What do you think is best?  Thanks!

Comment (Hamms) [reply bin/i18n/sync-codeorg-in.rb]: I'm more worried about the array values, which are keyed simply by index

Comment (joshlory) [reply bin/i18n/sync-codeorg-in.rb]: Ah, are you referring to the '0', '1', '2', '3'?  Those are specific enum values:

https://github.com/code-dot-org/code-dot-org/blob/54db780aa1f52da62ae15d336877e0c08f494bdf/dashboard/config/blocks/Dancelab/Dancelab_changeMove.json#L13-L33

Comment (Hamms) [reply bin/i18n/sync-codeorg-in.rb]: oh! I just assumed that was an array; my bad.

Even so, it does seem like there is potential for collision, albiet a small potential.

Comment (joshlory) [reply bin/i18n/sync-codeorg-in.rb]: Another option would be to prefix each option with the argument name.  Though I think that'll make it harder to parse back in at the other end.

```yaml
Dancelab_addBehavior:
  text: sprite {SPRITE} begins {BEHAVIOR}
Dancelab_atTimestamp:
  text: at {TIMESTAMP} seconds
Dancelab_changeColorBy:
  text: change {COLOR} {METHOD} by {AMOUNT}
  options:
    METHOD "hue": hue
    METHOD "saturation": saturation
    METHOD "brightness": brightness
Dancelab_changeMove:
  text: change {SPRITE} dance move to {MOVE}
  options:
    MOVE 0: Standing
    MOVE 1: Clap
    MOVE 2: Spin
    MOVE 3: Uprock
    MOVE 4: Bboy
    MOVE "next": Next
    MOVE "rand": Random
Dancelab_changePropBy:
  text: change {SPRITE} {PROPERTY} by {VAL}
  options:
    PROPERTY "scale": size
    PROPERTY "rotation": rotation
    PROPERTY "x": x position
    PROPERTY "y": y position
    PROPERTY "direction": movement direction
    PROPERTY "tint": tint
```

Comment (joshlory) [reply bin/i18n/sync-codeorg-in.rb]: What does the "redacted" version of this look like?  I was thinking `change {COLOR} {METHOD} by {AMOUNT}` would redact to `change {0} {1} by {2}` to make it clear that COLOR, METHOD, AMOUNT shouldn't be translated.  Does that match what you're thinking?

**Comment by joshlory** [comment]
Updated yaml format:

```yaml
---
en:
  data:
    blocks:
      Dancelab_mappingProp:
        text: "{PROPERTY} following {RANGE}"
        options:
          PROPERTY:
            '"scale"': size
            '"rotation"': rotation
            '"x"': x position
            '"y"': y position
            '"direction"': movement direction
            '"tint"': tint
          RANGE:
            '"bass"': bass
            '"mid"': mid
            '"treble"': treble
      gamelab_setPosition:
        text: move {THIS} to the {POSITION} position
        options:
          POSITION:
            "{x: 50, y: 50}": top left
            "{x: 200, y: 50}": top center
            "{x: 350, y: 50}": top right
            "{x: 50, y: 200}": center left
            "{x: 200, y: 200}": center
            "{x: 350, y: 200}": center right
            "{x: 50, y: 350}": bottom left
            "{x: 200, y: 350}": bottom center
            "{x: 350, y: 350}": bottom right
            '"random"': random
```

**Pull request merged**